core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -1281,6 +1284,81 @@ public MaterializedViewFreshness getMaterializedViewFreshness(Session session, Q
     }
 
+    @Override
+    public QualifiedObjectName getRedirectedTableName(Session session, QualifiedObjectName originalTableName)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(originalTableName, "originalTableName is null");
+
+        QualifiedObjectName tableName = originalTableName;
+        Set<QualifiedObjectName> visitedTableNames = new LinkedHashSet<>();
+        for (int count = 0; count < MAX_TABLE_REDIRECTIONS; count++) {
+            // Check for loop in redirection
+            if (!visitedTableNames.add(tableName)) {
+                throw new TrinoException(TABLE_REDIRECTION_ERROR,
+                        format("Table redirections form a loop: %s",
+                                getRedirectionChain(ImmutableList.<QualifiedObjectName>builder()
+                                        .addAll(visitedTableNames)
+                                        .add(tableName)
+                                        .build())));
+            }
+
+            Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, tableName.getCatalogName());
+            if (catalog.isPresent()) {
+                CatalogMetadata catalogMetadata = catalog.get();
+                CatalogName catalogName = catalogMetadata.getConnectorId(session, tableName);
+                ConnectorMetadata metadata = catalogMetadata.getMetadataFor(catalogName);
+
+                Optional<QualifiedObjectName> redirectedTableName = metadata.redirectTable(session.toConnectorSession(catalogName), tableName.asSchemaTableName())
+                        .map(name -> convertFromSchemaTableName(name.getCatalogName()).apply(name.getSchemaTableName()));
+
+                if (redirectedTableName.isEmpty()) {
+                    return tableName;
+                }
+
+                tableName = redirectedTableName.get();
+            }
+            else {
+                // Stop redirection
+                return tableName;
+            }
+        }
+        throw new TrinoException(TABLE_REDIRECTION_ERROR, format("Table redirected too many times (%d): %s", MAX_TABLE_REDIRECTIONS, visitedTableNames));
+    }
+
+    private static String getRedirectionChain(Collection<QualifiedObjectName> names)
+    {
+        return names.stream()
+                .map(QualifiedObjectName::toString)
+                .collect(Collectors.joining(" -> "));
+    }
+
+    @Override
+    public Optional<TableHandle> getRedirectedTableHandle(Session session, QualifiedObjectName tableName)
+    {
+        QualifiedObjectName targetTableName = getRedirectedTableName(session, tableName);
+
+        if (!targetTableName.equals(tableName)) {
+            Optional<TableHandle> tableHandle = getTableHandle(session, targetTableName);
+
+            if (tableHandle.isEmpty()) {
+                // Redirected table must exist
+                String redirectionMessage = format("Table '%s' has been redirected to '%s'.", tableName, targetTableName);
```

Comment:
As above, I think it's nicer to say "Table redirected to" which is present tense and explains the current state, rather than past tense.

From a grammar perspective, this might still be off, since "redirected" itself ending in "ed" is past tense, but the phrase "the table is redirected" sounds ok to me (as a native speaker who didn't pay much attention to such things in school) to describe the current state.